collaborative-plonk/src/round5.rs

```diff
@@ -0,0 +1,429 @@
+    round3::FinalPolys,
+    round4::{Round4Challenges, Round4Proof},
+    types::Keccak256Transcript,
+    Domains, FieldShare, FieldShareVec, PlonkData, PlonkProofResult,
+};
+use ark_ec::pairing::Pairing;
+use ark_ff::Field;
+use circom_types::{
+    plonk::PlonkProof,
+    traits::{CircomArkworksPairingBridge, CircomArkworksPrimeFieldBridge},
+};
+use mpc_core::traits::{
+    FFTPostProcessing, FFTProvider, MSMProvider, PairingEcMpcProtocol, PrimeFieldMpcProtocol,
+};
+use num_traits::One;
+use num_traits::Zero;
+
+pub(super) struct Round5<T, P: Pairing>
+where
+    for<'a> T: PrimeFieldMpcProtocol<P::ScalarField>
+        + PairingEcMpcProtocol<P>
+        + FFTProvider<P::ScalarField>
+        + MSMProvider<P::G1>
+        + MSMProvider<P::G2>,
+    P::ScalarField: mpc_core::traits::FFTPostProcessing,
+{
+    pub(super) driver: T,
+    pub(super) domains: Domains<P>,
+    pub(super) challenges: Round4Challenges<P>,
+    pub(super) proof: Round4Proof<P>,
+    pub(super) polys: FinalPolys<T, P>,
+    pub(super) data: PlonkData<T, P>,
+}
+pub(super) struct Round5Challenges<P: Pairing> {
+    beta: P::ScalarField,
+    gamma: P::ScalarField,
+    alpha: P::ScalarField,
+    xi: P::ScalarField,
+    v: [P::ScalarField; 5],
+}
+
+impl<P: Pairing> Round5Challenges<P> {
+    fn new(round4_challenges: Round4Challenges<P>, v: [P::ScalarField; 5]) -> Self {
+        Self {
+            beta: round4_challenges.beta,
+            gamma: round4_challenges.gamma,
+            alpha: round4_challenges.alpha,
+            xi: round4_challenges.xi,
+            v,
+        }
+    }
+}
+
+impl<P: Pairing + CircomArkworksPairingBridge> Round4Proof<P>
+where
+    P::BaseField: CircomArkworksPrimeFieldBridge,
+    P::ScalarField: CircomArkworksPrimeFieldBridge,
+{
+    fn into_final_proof(self, commit_wxi: P::G1, commit_wxiw: P::G1) -> PlonkProof<P> {
+        PlonkProof {
+            a: self.commit_a.into(),
+            b: self.commit_b.into(),
+            c: self.commit_c.into(),
+            z: self.commit_z.into(),
+            t1: self.commit_t1.into(),
+            t2: self.commit_t2.into(),
+            t3: self.commit_t3.into(),
+            eval_a: self.eval_a,
+            eval_b: self.eval_b,
+            eval_c: self.eval_c,
+            eval_s1: self.eval_s1,
+            eval_s2: self.eval_s2,
+            eval_zw: self.eval_zw,
+            wxi: commit_wxi.into(),
+            wxiw: commit_wxiw.into(),
+            protocol: "plonk".to_string(),
+            curve: P::get_circom_name(),
+        }
+    }
+}
+
+impl<T, P: Pairing> Round5<T, P>
+where
+    for<'a> T: PrimeFieldMpcProtocol<P::ScalarField>
+        + PairingEcMpcProtocol<P>
+        + FFTProvider<P::ScalarField>
+        + MSMProvider<P::G1>
+        + MSMProvider<P::G2>,
+    P::ScalarField: FFTPostProcessing,
+    P: CircomArkworksPairingBridge,
+    P::BaseField: CircomArkworksPrimeFieldBridge,
+    P::ScalarField: CircomArkworksPrimeFieldBridge,
+{
+    fn div_by_zerofier(
```

Comment:
Is `zerofier` a word?

Reply:
Found it in papers as well, e.g., https://eprint.iacr.org/2020/1022.pdf